src/libtiled/varianttomapconverter.cpp

```diff
@@ -334,20 +342,51 @@ TemplateGroup *VariantToMapConverter::toTemplateGroup(const QVariant &variant)
 {
     const QVariantMap variantMap = variant.toMap();
 
+    const QVariant sourceVariant = variantMap[QLatin1String("source")];
     const int firstTid = variantMap[QLatin1String("firsttid")].toInt();
+    const int nextTemplateId = variantMap[QLatin1String("nexttemplateid")].toInt();
+    const QString name = variantMap[QLatin1String("name")].toString();
 
-    const QVariant sourceVariant = variantMap[QLatin1String("source")];
-    QString source = resolvePath(mMapDir, sourceVariant);
-    QString error;
-    TemplateGroup *templateGroup = TemplateManager::instance()->loadTemplateGroup(source, &error);
-    if (!templateGroup) { // Couldn't open the file
-        templateGroup = new TemplateGroup();
-        templateGroup->setFileName(source);
-        templateGroup->setLoaded(false);
+    TemplateGroup *templateGroup;
+
+    if (sourceVariant.isNull()) {
+        templateGroup = new TemplateGroup(name);
+
+        templateGroup->setNextTemplateId(nextTemplateId);
+
+        const auto tilesetVariants = variantMap[QLatin1String("tilesets")].toList();
+        for (const QVariant &tilesetVariant : tilesetVariants) {
+            SharedTileset tileset = toTileset(tilesetVariant);
+            if (!tileset)
+                return nullptr;
+
+            templateGroup->addTileset(tileset);
+        }
+
+        const auto templateVariants = variantMap[QLatin1String("templates")].toList();
+        for (const QVariant &templateVariant : templateVariants) {
+            ObjectTemplate *objectTemplate = toObjectTemplate(templateVariant.toMap());
+            if (!objectTemplate)
+                return nullptr;
+
+            templateGroup->addTemplate(objectTemplate);
+        }
     } else {
-        mTidMapper.insert(firstTid, templateGroup);
+        QString source = resolvePath(mMapDir, sourceVariant);
+
+        QString error;
+        templateGroup = TemplateManager::instance()->loadTemplateGroup(source, &error);
+
+        if (!templateGroup) { // Couldn't open the file
+            templateGroup = new TemplateGroup();
+            templateGroup->setFileName(source);
+            templateGroup->setLoaded(false);
+        }
     }
 
+    if (templateGroup)
```

Comment:
This check is superfluous since the above code always makes sure there is a template group.